MFVNeutralino/plugins/Vertexer.cc

```diff
@@ -695,6 +737,189 @@ void MFVVertexer::produce(edm::Event& event, const edm::EventSetup& setup) {
+        std::cout << std::endl;
+      }
+      useful_tracks = false;
+      for (const reco::TrackRef& itk:track_discard){
+        const reco::TransientTrack& ttk = seed_tracks[seed_track_ref_map[itk]];
+        int v_assign = -1;
+        double v_assign_dist_sig = 999;
+        unsigned int v_assign_ntk = 0;
+        if (verbose){
+          std::cout << "For track " << itk.key() << std::endl;
+        }
+        for (size_t i=0; i<vertices->size(); ++i){
+          const reco::Vertex& v = vertices->at(i);
+          std::pair<bool, Measurement1D> t_dist = track_dist(ttk, v);
+          t_dist.first = t_dist.first && (t_dist.second.value() < max_track_vertex_dist || t_dist.second.significance() < max_track_vertex_sig); // whether it is too far away from the vtx
+          if (t_dist.first) {
+            if (t_dist.second.significance()<min_track_vertex_sig_to_remove){
+              if (v_assign_ntk<v.nTracks()){
+                v_assign = i;
+                v_assign_dist_sig = t_dist.second.significance();
+                v_assign_ntk = v.nTracks();
+              }
+            }
+            else if (t_dist.second.significance()<v_assign_dist_sig){
+              v_assign = i;
+              v_assign_dist_sig = t_dist.second.significance();
+              v_assign_ntk = v.nTracks();
+            }
+          }
+          
+          if (verbose){
+            std::cout << "  Track-vertex " << i << " " << t_dist.first << " dist: " << t_dist.second.value() << " sig: " << t_dist.second.significance() << std::endl;
+          }
+        }
+        if (verbose)
+          std::cout << "Track " << itk.key() << " assigned to " << v_assign << std::endl;
+        if (v_assign>=0){
+          useful_tracks = true;
+          track_discard.erase(itk);
+          if (verbose){
+            std::cout << "Fitting vertex " << v_assign << " with tracks: ";
+            print_track_set((*vertices)[v_assign]);
+            std::cout << " with added track " << itk.key() << std::endl;
+          }
+          std::vector<reco::TransientTrack> ttks;
+          for (auto tk:vertex_track_set((*vertices)[v_assign])){
+            ttks.push_back(seed_tracks[seed_track_ref_map[tk]]);
+          }
+          ttks.push_back(ttk);
+          reco::VertexCollection new_vertices;
+          for (const TransientVertex& tv : kv_reco_dropin(ttks))
+            new_vertices.push_back(reco::Vertex(tv));
+          if (verbose) {
+            printf("      got %lu new vertices out of the av fit for v%i\n", new_vertices.size(), v_assign);
+            printf("      these track sets:");
+            for (const auto& nv : new_vertices) {
+              printf(" (");
+              print_track_set(nv);
+              printf(" ),");
+            }
+            printf("\n");
+          }
+          if (new_vertices.size() == 1)
+            (*vertices)[v_assign] = new_vertices[0];
+          break;
+        }
+      }
+    }
+  }
+
+  // track attachment
```

Comment:
This idea came from the ATLAS paper, right? Maybe adding a link could be nice, for when we forget this!